### PR TITLE
[LI-HOTFIX] Handle any type of Exception during log recovery

### PR DESCRIPTION
**Pull request by gitlw**
TICKET = [LIKAFKA-47870](https://jira01.corp.linkedin.com:8443/browse/LIKAFKA-47870)
LI_DESCRIPTION =
In the current implementation, only the InvalidOffsetException is handled during log recovery.
This PR changes the implementation so that any other types of Exception would also cause a log truncation,
and hence not block the broker startup.

We are also adding tests to verify that brokers can successfully startup after corruption is found in the size field of a batch, or an unexpected exception is thrown.
EXIT_CRITERIA = N/A

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/log/LogLoader.scala [hunk 1]
```diff
@@ -351,6 +350,12 @@ object LogLoader extends Logging {
    * @throws LogSegmentOffsetOverflowException if the segment contains messages that cause index offset overflow
    */
   private def recoverSegment(segment: LogSegment, params: LoadLogParams): Int = {
+    if (GlobalConfig.logRecoveryShouldThrowException) {
+      // we should only trigger the exception once
+      GlobalConfig.logRecoveryShouldThrowException = false
+      throw new IllegalStateException("Explicitly trigger an exception to test the log recovery logic")
```
Comment (groelofs) [inline core/src/main/scala/kafka/log/LogLoader.scala:356]: Did you intend to leave this block in here? Seems like a temporary/testing item, even if it's protected by a config.

File: core/src/main/scala/kafka/log/LogLoader.scala [hunk 2]
```diff
@@ -418,10 +423,12 @@ object LogLoader extends Logging {
           try {
             recoverSegment(segment, params)
           } catch {
-            case _: InvalidOffsetException =>
+            case ooe: LogSegmentOffsetOverflowException =>
```
Comment (groelofs) [inline core/src/main/scala/kafka/log/LogLoader.scala:426]: Are there any other exception types that should be omitted? Obviously truncation = data loss = a fairly permanent, fairly serious action to take. Might there be some transient ones (e.g., perhaps some types of IOException at the JVM layer, perhaps converted to RuntimeException before getting here) that would go away on retry but would result in truncation with this new change?

I agree with the general intent here, but the approach is a bit heavy-handed, so I'm worried about unintended consequences.

File: core/src/main/scala/kafka/server/GlobalConfig.scala [hunk 3]
```diff
@@ -20,4 +20,5 @@ package kafka.server
 object GlobalConfig {
   // a global flag to indicate whether the corrupted files should be dropped
   var liDropCorruptedFilesEnable = false
+  @volatile var logRecoveryShouldThrowException = false
```
Comment (groelofs) [inline core/src/main/scala/kafka/server/GlobalConfig.scala:24]: Please add a comment describing the intent behind this, any warnings/concerns about its use in prod, and its intended lifecycle (if it's a short-term hack).

File: core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala [hunk 4]
```diff
@@ -183,5 +197,118 @@ class DropCorruptedFilesTest extends ZooKeeperTestHarness {
     bw.close()
   }
 
+  private def setupSingleBrokerClusterWithTopic(segmentBytes: Long): (KafkaServer, KafkaProducer[Array[Byte], Array[Byte]]) = {
+    // create brokers
+    val serverConfigs = TestUtils.createBrokerConfigs(2, zkConnect, false)
```
Comment (groelofs) [inline core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala:202]: nit: 2 configs for a "SingleBroker"?  This is a DualBrokerClusterWithTopic, right?

File: core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala [hunk 5]
```diff
@@ -183,5 +197,118 @@ class DropCorruptedFilesTest extends ZooKeeperTestHarness {
...
+
+    // delete the recovery-point-offset-checkpoint file so that all log segments need to go through recovery
+    val recoveryPointCheckpointFile = new File(logDir, RecoveryPointCheckpointFile)
+    recoveryPointCheckpointFile.delete()
+
+    dataBroker.replicaManager.logManager
+    dataBroker.startup()
+
+    // verify that the segments after the corrupted are deleted and the corrupted segment is truncated to size 0
+    assertEquals(2, segmentsOnBroker.numberOfSegments)
```
Comment (groelofs) [inline core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala:268]: Is there a race condition here with dataBroker.startup(), or is the latter a synchronous call?

File: core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala [hunk 6]
```diff
@@ -183,5 +197,118 @@ class DropCorruptedFilesTest extends ZooKeeperTestHarness {
...
+    // delete the recovery-point-offset-checkpoint file so that all log segments need to go through recovery
+    val recoveryPointCheckpointFile = new File(logDir, RecoveryPointCheckpointFile)
+    recoveryPointCheckpointFile.delete()
+
+    GlobalConfig.logRecoveryShouldThrowException = true
+    dataBroker.replicaManager.logManager
+    dataBroker.startup()
+
+    // verify that only the active segment is left with a size of 0
+    assertEquals(1, segmentsOnBroker.numberOfSegments)
```
Comment (groelofs) [inline core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala:303]: Same race-condition question.

File: core/src/main/scala/kafka/log/LogLoader.scala (hunk 1 above)
Comment (gitlw) [reply core/src/main/scala/kafka/log/LogLoader.scala:356]: Yes, I indent to leave this block here since it's necessary for the test.

File: core/src/main/scala/kafka/log/LogLoader.scala (hunk 2 above)
Comment (gitlw) [reply core/src/main/scala/kafka/log/LogLoader.scala:426]: Your concern is valid. I've changed the logic so that the log truncation will only happen when it's InvalidOffsetException or when the recovery flag is turned on.

File: core/src/main/scala/kafka/server/GlobalConfig.scala (hunk 3 above)
Comment (gitlw) [reply core/src/main/scala/kafka/server/GlobalConfig.scala:24]: Done

File: core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala (hunk 4 above)
Comment (gitlw) [reply core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala:202]: added clarification that "SingleBroker" here means "SingleDataBroker" not including the controller.

File: core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala (hunk 5 above)
Comment (gitlw) [reply core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala:268]: Yes, the startup() method is a synchroneous call.

File: core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala (hunk 6 above)
Comment (gitlw) [reply core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala:303]: ditto

File: core/src/main/scala/kafka/server/GlobalConfig.scala [hunk 7]
```diff
@@ -20,5 +20,6 @@ package kafka.server
 object GlobalConfig {
   // a global flag to indicate whether the corrupted files should be dropped
   var liDropCorruptedFilesEnable = false
+  // The logRecoveryShouldThrowException should only be used for testing purposes. Never turn in on in a real cluster!
```
Comment (groelofs) [inline core/src/main/scala/kafka/server/GlobalConfig.scala]: nit: "turn _it_ on in ..."

**Review by groelofs — APPROVED** [review]
Looks good, thanks! Just a single-letter typo left.
